File: evaluation/common/memory_pipeline/validation.py
"""Schema, evidence, and deterministic guards for extracted memories."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping, Sequence

from .canonical import stable_hash
from .models import (
    AtomicMemory,
    EvidenceRef,
    ExtractionWindow,
    MessageRef,
    NormalizedMessage,
    PipelineIssue,
)


MEMORY_TYPES = {
    "fact",
    "preference",
    "relationship",
    "event",
    "state",
    "procedure",
    "other",
}
MODALITIES = {
    "asserted",
    "negated",
    "possible",
    "planned",
    "conditional",
    "reported",
}
EVIDENCE_ROLES = {"primary", "supporting"}
_PLAN_MARKERS = (" plan ", "plan to", "intend to", "打算", "计划", "准备")
_POSSIBLE_MARKERS = ("might", "maybe", "possibly", "可能", "也许", "或许")
_NEGATION_MARKERS = (" don't ", " not ", "never", "不", "没", "从不")


@dataclass(frozen=True)
class ValidationConfig:
    max_memory_chars: int = 500

    def __post_init__(self) -> None:
        if self.max_memory_chars <= 0:
            raise ValueError("max_memory_chars must be positive")


@dataclass(frozen=True)
class ValidationBatch:
    valid: list[AtomicMemory]
    rejected: list[PipelineIssue]
    quarantined: list[PipelineIssue]


def validate_extraction(
    raw_memories: Sequence[dict[str, Any]],
    window: ExtractionWindow,
    messages_by_id: Mapping[str, NormalizedMessage],
    config: ValidationConfig,
) -> ValidationBatch:
    valid: list[AtomicMemory] = []
    rejected: list[PipelineIssue] = []
    quarantined: list[PipelineIssue] = []

    for index, raw in enumerate(raw_memories):
        schema_issue = _schema_issue(raw, window, index)
        if schema_issue is not None:
            rejected.append(schema_issue)
            continue

        evidence, evidence_issue, issue_kind = _validate_evidence(
            raw["evidence"],
            window,
            messages_by_id,
            index,
        )
        if evidence_issue is not None:
            if issue_kind == "rejected":
                rejected.append(evidence_issue)
            else:
                quarantined.append(evidence_issue)
            continue

        if not any(
            item.evidence_role == "primary"
            and _is_candidate_span(item, window.candidate_refs)
            for item in evidence
        ):
            rejected.append(
                _issue(
                    window,
                    index,
                    "missing_candidate_evidence",
                    "memory requires primary evidence from a candidate span",
                )
            )
            continue

        if len(raw["text"]) > config.max_memory_chars:
            quarantined.append(
                _issue(
                    window,
                    index,
                    "memory_text_too_long",
                    f"memory text exceeds {config.max_memory_chars} characters",
                )
            )
            continue

        if _suspicious_modality(raw["modality"], evidence):
            quarantined.append(
                _issue(
                    window,
                    index,
                    "suspicious_modality",
                    "memory modality is inconsistent with its evidence",
                )
            )
            continue

        canonical_content = {
            "memory_type": raw["memory_type"],
            "text": raw["text"].strip(),
            "subject": dict(raw["subject"]),
            "predicate": raw["predicate"].strip(),
            "object": raw.get("object"),
            "modality": raw["modality"],
            "event_time": dict(raw["event_time"]) if raw.get("event_time") else None,
            "attributes": dict(raw.get("attributes") or {}),
        }
        valid.append(
            AtomicMemory(
                id="candidate-" + stable_hash(window.scope_id, canonical_content),
                scope_id=window.scope_id,
                evidence=tuple(evidence),
                model_confidence=raw.get("model_confidence"),
                observed_at=_observed_at(evidence, messages_by_id, window),
                source_episode_id=window.episode_id,
                source_window_id=window.id,
                **canonical_content,
            )
        )

    return ValidationBatch(valid=valid, rejected=rejected, quarantined=quarantined)


def _schema_issue(
    raw: Any,
    window: ExtractionWindow,
    index: int,
) -> PipelineIssue | None:
    if not isinstance(raw, dict):
        return _issue(window, index, "malformed_schema", "memory must be an object")

    text = raw.get("text")
    subject = raw.get("subject")
    predicate = raw.get("predicate")
    object_value = raw.get("object")
    evidence = raw.get("evidence")
    attributes = raw.get("attributes", {})
    event_time = raw.get("event_time")
    confidence = raw.get("model_confidence")
    if (
        not isinstance(text, str)
        or not text.strip()
        or not isinstance(subject, dict)
        or not isinstance(predicate, str)
        or not predicate.strip()
        or not (
            object_value is None
            or isinstance(object_value, (str, dict))
        )
        or not isinstance(evidence, list)
        or not evidence
        or not isinstance(attributes, dict)
        or (event_time is not None and not isinstance(event_time, dict))
        or (
            confidence is not None
            and (
                not isinstance(confidence, (int, float))
                or isinstance(confidence, bool)
                or not 0.0 <= float(confidence) <= 1.0
            )
        )
    ):
        return _issue(
            window,
            index,
            "malformed_schema",
            "memory fields do not match atomic_memory_v1",
        )

    if raw.get("memory_type") not in MEMORY_TYPES or raw.get("modality") not in MODALITIES:
        return _issue(
            window,
            index,
            "unknown_enum",
            "memory_type or modality is not allowed",
        )
    return None


def _validate_evidence(
    raw_evidence: Sequence[dict[str, Any]],
    window: ExtractionWindow,
    messages_by_id: Mapping[str, NormalizedMessage],
    index: int,
) -> tuple[list[EvidenceRef], PipelineIssue | None, str]:
    window_refs = (
        *window.context_before_refs,
        *window.candidate_refs,
        *window.context_after_refs,
    )
    output: list[EvidenceRef] = []
    for raw in raw_evidence:
        if not isinstance(raw, dict):
            return [], _issue(window, index, "malformed_schema", "evidence must be an object"), "rejected"
        message_id = raw.get("message_id")
        quote = raw.get("quote")
        role = raw.get("evidence_role")
        if (
            not isinstance(message_id, str)
            or not isinstance(quote, str)
            or not quote
            or role not in EVIDENCE_ROLES
        ):
            return [], _issue(window, index, "malformed_schema", "evidence fields are invalid"), "rejected"

        refs = [ref for ref in window_refs if ref.message_id == message_id]
        if not refs or message_id not in messages_by_id:
            return [], _issue(
                window,
                index,
                "unknown_evidence_message",
                f"evidence message {message_id!r} is not in the window",
            ), "rejected"

        matches: list[tuple[MessageRef, int]] = []
        for ref in refs:
            search_at = 0
            while True:
                position = ref.text.find(quote, search_at)
                if position < 0:
                    break
                matches.append((ref, position))
                search_at = position + 1
        if not matches:
            return [], _issue(
                window,
                index,
                "evidence_quote_not_found",
                "evidence quote is not an exact substring of the referenced window span",
            ), "quarantine"
        if len(matches) != 1:
            return [], _issue(
                window,
                index,
                "ambiguous_evidence_quote",
                "evidence quote occurs more than once in the referenced window spans",
            ), "quarantine"

        ref, local_start = matches[0]
        start_char = ref.start_char + local_start
        output.append(
            EvidenceRef(
                message_id=message_id,
                quote=quote,
                start_char=start_char,
                end_char=start_char + len(quote),
                evidence_role=role,
            )
        )
    return output, None, ""


def _is_candidate_span(
    evidence: EvidenceRef,
    candidates: Sequence[MessageRef],
) -> bool:
    return any(
        ref.message_id == evidence.message_id
        and ref.start_char <= evidence.start_char
        and evidence.end_char <= ref.end_char
        for ref in candidates
    )


def _suspicious_modality(
    modality: str,
    evidence: Sequence[EvidenceRef],
) -> bool:
    if modality != "asserted":
        return False
    text = " " + " ".join(item.quote for item in evidence).lower() + " "
    return any(marker in text for marker in (*_PLAN_MARKERS, *_POSSIBLE_MARKERS, *_NEGATION_MARKERS))


def _observed_at(
    evidence: Sequence[EvidenceRef],
    messages_by_id: Mapping[str, NormalizedMessage],
    window: ExtractionWindow,
) -> str:
    for item in reversed(evidence):
        timestamp = messages_by_id[item.message_id].timestamp
        if timestamp:
            return timestamp
    for ref in reversed(window.candidate_refs):
        timestamp = messages_by_id[ref.message_id].timestamp
        if timestamp:
            return timestamp
    return ""


def _issue(
    window: ExtractionWindow,
    index: int,
    code: str,
    message: str,
) -> PipelineIssue:
    return PipelineIssue(
        stage="validation",
        code=code,
        message=message,
        scope_id=window.scope_id,
        episode_id=window.episode_id,
        window_id=window.id,
        details={"memory_index": index},
    )

File: crates/memory-pipeline/src/validation.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};

use crate::canonical::stable_hash;
use crate::models::{
    AtomicMemory, EvidenceRef, ExtractionWindow, MessageRef, NormalizedMessage, PipelineIssue,
};

const MEMORY_TYPES: [&str; 7] = [
    "fact",
    "preference",
    "relationship",
    "event",
    "state",
    "procedure",
    "other",
];
const MODALITIES: [&str; 6] = [
    "asserted",
    "negated",
    "possible",
    "planned",
    "conditional",
    "reported",
];
const EVIDENCE_ROLES: [&str; 2] = ["primary", "supporting"];

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct ValidationConfig {
    pub max_memory_chars: usize,
}

impl Default for ValidationConfig {
    fn default() -> Self {
        Self {
            max_memory_chars: 500,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ValidationBatch {
    pub valid: Vec<AtomicMemory>,
    pub rejected: Vec<PipelineIssue>,
    pub quarantined: Vec<PipelineIssue>,
}

pub fn validate_extraction(
    raw_memories: &[Value],
    window: &ExtractionWindow,
    messages_by_id: &HashMap<String, NormalizedMessage>,
    config: &ValidationConfig,
) -> ValidationBatch {
    let mut batch = ValidationBatch::default();
    for (index, raw) in raw_memories.iter().enumerate() {
        let Some(object) = valid_schema(raw) else {
            batch.rejected.push(issue(
                window,
                index,
                "malformed_schema",
                "memory fields do not match atomic_memory_v1",
            ));
            continue;
        };
        let memory_type = object["memory_type"].as_str().unwrap();
        let modality = object["modality"].as_str().unwrap();
        if !MEMORY_TYPES.contains(&memory_type) || !MODALITIES.contains(&modality) {
            batch.rejected.push(issue(
                window,
                index,
                "unknown_enum",
                "memory_type or modality is not allowed",
            ));
            continue;
        }
        let (evidence, evidence_issue, rejected) = validate_evidence(
            object["evidence"].as_array().unwrap(),
            window,
            messages_by_id,
            index,
        );
        if let Some(problem) = evidence_issue {
            if rejected {
                batch.rejected.push(problem)
            } else {
                batch.quarantined.push(problem)
            }
            continue;
        }
        if !evidence.iter().any(|item| {
            item.evidence_role == "primary" && is_candidate(item, &window.candidate_refs)
        }) {
            batch.rejected.push(issue(
                window,
                index,
                "missing_candidate_evidence",
                "memory requires primary evidence from a candidate span",
            ));
            continue;
        }
        let memory_text = object["text"].as_str().unwrap().trim();
        if memory_text.chars().count() > config.max_memory_chars {
            batch.quarantined.push(issue(
                window,
                index,
                "memory_text_too_long",
                &format!("memory text exceeds {} characters", config.max_memory_chars),
            ));
            continue;
        }
        if suspicious_modality(modality, &evidence) {
            batch.quarantined.push(issue(
                window,
                index,
                "suspicious_modality",
                "memory modality is inconsistent with its evidence",
            ));
            continue;
        }
        let subject = object["subject"].as_object().unwrap().clone();
        let event_time = object
            .get("event_time")
            .and_then(Value::as_object)
            .filter(|value| !value.is_empty())
            .cloned();
        let attributes = object
            .get("attributes")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        let canonical = json!({
            "memory_type": memory_type,
            "text": memory_text,
            "subject": subject,
            "predicate": object["predicate"].as_str().unwrap().trim(),
            "object": object.get("object").cloned().unwrap_or(Value::Null),
            "modality": modality,
            "event_time": event_time,
            "attributes": attributes,
        });
        let observed_at = observed_at(&evidence, messages_by_id, window);
        batch.valid.push(AtomicMemory {
            id: format!(
                "candidate-{}",
                stable_hash(&[json!(window.scope_id), canonical.clone()])
            ),
            scope_id: window.scope_id.clone(),
            text: memory_text.into(),
            memory_type: memory_type.into(),
            subject,
            predicate: object["predicate"].as_str().unwrap().trim().into(),
            object: object
                .get("object")
                .filter(|value| !value.is_null())
                .cloned(),
            modality: modality.into(),
            evidence,
            event_time,
            attributes,
            model_confidence: object.get("model_confidence").cloned(),
            observed_at,
            source_episode_id: window.episode_id.clone(),
            source_window_id: window.id.clone(),
            observation_refs: Vec::new(),
        });
    }
    batch
}

fn valid_schema(raw: &Value) -> Option<&Map<String, Value>> {
    let object = raw.as_object()?;
    let text = object.get("text")?.as_str()?;
    let predicate = object.get("predicate")?.as_str()?;
    if text.trim().is_empty() || predicate.trim().is_empty() || !object.get("subject")?.is_object()
    {
        return None;
    }
    let object_value = object.get("object").unwrap_or(&Value::Null);
    if !(object_value.is_null() || object_value.is_string() || object_value.is_object()) {
        return None;
    }
    if object
        .get("evidence")?
        .as_array()
        .is_none_or(|items| items.is_empty())
    {
        return None;
    }
    if !object.get("attributes").unwrap_or(&json!({})).is_object() {
        return None;
    }
    if object
        .get("event_time")
        .is_some_and(|value| !(value.is_null() || value.is_object()))
    {
        return None;
    }
    if let Some(confidence) = object.get("model_confidence") {
        let value = confidence.as_f64()?;
        if !(0.0..=1.0).contains(&value) {
            return None;
        }
    }
    object.get("memory_type")?.as_str()?;
    object.get("modality")?.as_str()?;
    Some(object)
}

fn validate_evidence(
    raw: &[Value],
    window: &ExtractionWindow,
    messages: &HashMap<String, NormalizedMessage>,
    index: usize,
) -> (Vec<EvidenceRef>, Option<PipelineIssue>, bool) {
    let refs = window
        .context_before_refs
        .iter()
        .chain(window.candidate_refs.iter())
        .chain(window.context_after_refs.iter())
        .collect::<Vec<_>>();
    let mut output = Vec::new();
    for value in raw {
        let Some(value) = value.as_object() else {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "malformed_schema",
                    "evidence must be an object",
                )),
                true,
            );
        };
        let (Some(message_id), Some(quote), Some(role)) = (
            value.get("message_id").and_then(Value::as_str),
            value.get("quote").and_then(Value::as_str),
            value.get("evidence_role").and_then(Value::as_str),
        ) else {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "malformed_schema",
                    "evidence fields are invalid",
                )),
                true,
            );
        };
        if quote.is_empty() || !EVIDENCE_ROLES.contains(&role) {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "malformed_schema",
                    "evidence fields are invalid",
                )),
                true,
            );
        }
        let matching_refs = refs
            .iter()
            .filter(|reference| reference.message_id == message_id)
            .collect::<Vec<_>>();
        if matching_refs.is_empty() || !messages.contains_key(message_id) {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "unknown_evidence_message",
                    &format!("evidence message {message_id:?} is not in the window"),
                )),
                true,
            );
        }
        let mut matches = Vec::new();
        for reference in matching_refs {
            let source = reference.text.chars().collect::<Vec<_>>();
            let needle = quote.chars().collect::<Vec<_>>();
            if needle.len() <= source.len() {
                for local_char in 0..=source.len() - needle.len() {
                    if source[local_char..local_char + needle.len()] == needle {
                        matches.push((reference, local_char));
                    }
                }
            }
        }
        if matches.is_empty() {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "evidence_quote_not_found",
                    "evidence quote is not an exact substring of the referenced window span",
                )),
                false,
            );
        }
        if matches.len() != 1 {
            return (
                Vec::new(),
                Some(issue(
                    window,
                    index,
                    "ambiguous_evidence_quote",
                    "evidence quote occurs more than once in the referenced window spans",
                )),
                false,
            );
        }
        let (reference, local_start) = matches[0];
        let start = reference.start_char + local_start;
        output.push(EvidenceRef {
            message_id: message_id.into(),
            quote: quote.into(),
            start_char: start,
            end_char: start + quote.chars().count(),
            evidence_role: role.into(),
        });
    }
    (output, None, false)
}

fn is_candidate(evidence: &EvidenceRef, refs: &[MessageRef]) -> bool {
    refs.iter().any(|reference| {
        reference.message_id == evidence.message_id
            && reference.start_char <= evidence.start_char
            && evidence.end_char <= reference.end_char
    })
}

fn suspicious_modality(modality: &str, evidence: &[EvidenceRef]) -> bool {
    if modality != "asserted" {
        return false;
    }
    let text = format!(
        " {} ",
        evidence
            .iter()
            .map(|item| item.quote.as_str())
            .collect::<Vec<_>>()
            .join(" ")
            .to_lowercase()
    );
    [
        " plan ",
        "plan to",
        "intend to",
        "打算",
        "计划",
        "准备",
        "might",
        "maybe",
        "possibly",
        "可能",
        "也许",
        "或许",
        " don't ",
        " not ",
        "never",
        "不",
        "没",
        "从不",
    ]
    .iter()
    .any(|marker| text.contains(marker))
}

fn observed_at(
    evidence: &[EvidenceRef],
    messages: &HashMap<String, NormalizedMessage>,
    window: &ExtractionWindow,
) -> String {
    if let Some(timestamp) = evidence.iter().rev().find_map(|item| {
        let timestamp = &messages.get(&item.message_id)?.timestamp;
        (!timestamp.is_empty()).then(|| timestamp.clone())
    }) {
        return timestamp;
    }
    window
        .candidate_refs
        .iter()
        .rev()
        .find_map(|reference| {
            let timestamp = &messages.get(&reference.message_id)?.timestamp;
            (!timestamp.is_empty()).then(|| timestamp.clone())
        })
        .unwrap_or_default()
}

fn issue(window: &ExtractionWindow, index: usize, code: &str, message: &str) -> PipelineIssue {
    PipelineIssue {
        stage: "validation".into(),
        code: code.into(),
        message: message.into(),
        scope_id: window.scope_id.clone(),
        episode_id: window.episode_id.clone(),
        window_id: window.id.clone(),
        details: Map::from_iter([("memory_index".into(), json!(index))]),
        ..PipelineIssue::default()
    }
}
